futures-util/src/stream/for_each_concurrent.rs

```diff
@@ -0,0 +1,106 @@
+use core::marker::Unpin;
+use core::mem::PinMut;
+use core::num::NonZeroUsize;
+use futures_core::future::Future;
+use futures_core::stream::Stream;
+use futures_core::task::{self, Poll};
+use pin_utils::{unsafe_pinned, unsafe_unpinned};
+use crate::stream::{FuturesUnordered, StreamExt};
+
+/// A stream combinator which executes a unit closure over each item on a
+/// stream concurrently.
+///
+/// This structure is returned by the
+/// [`StreamExt::for_each_concurrent`](super::StreamExt::for_each_concurrent)
+/// method.
+#[derive(Debug)]
+#[must_use = "streams do nothing unless polled"]
+pub struct ForEachConcurrent<St, Fut, F> {
+    stream: Option<St>,
+    f: F,
+    futures: FuturesUnordered<Fut>,
+    limit: Option<NonZeroUsize>,
+}
+
+impl<St, Fut, F> Unpin for ForEachConcurrent<St, Fut, F>
+where St: Unpin,
+      Fut: Unpin,
+{}
+
+impl<St, Fut, F> ForEachConcurrent<St, Fut, F>
+where St: Stream,
+      F: FnMut(St::Item) -> Fut,
+      Fut: Future<Output = ()>,
+{
+    unsafe_pinned!(stream: Option<St>);
+    unsafe_unpinned!(f: F);
+    unsafe_unpinned!(futures: FuturesUnordered<Fut>);
+    unsafe_unpinned!(limit: Option<NonZeroUsize>);
+
+    pub(super) fn new(stream: St, limit: Option<usize>, f: F) -> ForEachConcurrent<St, Fut, F> {
+        ForEachConcurrent {
+            stream: Some(stream),
+            // Note: `limit` = 0 gets ignored.
+            limit: limit.and_then(NonZeroUsize::new),
```

Comment:
I'm guessing there's no ergonomic way to push this up to the public API? What do you think of panicking here on `limit == 0` and calling it developer error instead?

Reply:
I suppose we could panic. I have no strong preference here. I think this is unlikely to be an issue in practice, so either way is fine.

Reply:
This should be consistent with `Buffered`.

I think that the making it possible to impose no limit is a good idea. limit = 0 should produce a stream that does nothing because it can't poll any futures. Not the best option, but IMO better than the alternatives:
- No limit is unintuitive because that's what `None` is for.
- A panic feels a bit drastic and is also not quite logical
- Exposing `NonZeroUsize` is a bit cumbersome

Reply:
I agree we should be consistent-- I think we should change `Buffered`/`BuffereUnordered` as well. IMO the size savings from `Option<NonZeroUsize` are worth it, and passing in zero to any of those functions will always be buggy / bad.

Reply:
Hmm actually it's quite common that `0` means no limit. I suppose that's also fine. However it should be mentioned in the docs for `for_each_concurrent`

Reply:
Done.